### PR TITLE
Write the unit test

### PR DESCRIPTION
**Pull request by axcell-0**
In this project we implemented the following changes.

Added unit tests for all the methods in all the classes:
Book
Person
Student
Teacher
Classroom
Rental
all classes that you created while implementing the Decorator design pattern.

### CODE REVIEW

File: spec/book_spec.rb [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+    it 'returns the title' do
+      @book.title.should eql 'House of Terror'
+    end
+  end
+
+  describe '#author' do
+    it 'returns the author' do
+      @book.author.should eql 'Steve Ford'
+    end
+  end
```
Comment (danmainah) [inline spec/book_spec.rb]: - [x] @Team Great work adding tests for the book classes, however you forgot to add tests for these methods ` add_rental, book_to_json, self.from_json, save_books and load_books`, its  project requirement to add tests for all methods
image

File: spec/person_spec.rb [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+  describe '#can_use_services' do
+    it 'should return true if of age' do
+      expect(person.can_use_services).to be(true)
+    end
+
+    it 'should return true if parent permission is given' do
+      person_with_permission = Person.new(17, 'Jane Doe', parent_permission: true)
+      expect(person_with_permission.can_use_services).to be(true)
+    end
+  end
```
Comment (danmainah) [inline spec/person_spec.rb:48]: - [ ] @Team Great work adding tests for the person class, however you forgot to add tests for these methods ` people_to_json, self.from_json, add_rentals, self.find_by_id,  save_people and load_people`, its  project requirement to add tests for all methods
image

File: spec/teacher_spec.rb [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+    end
+
+    it 'returns the name' do
+      expect(@teacher.name).to eql 'Bianca'
+    end
+
+    it 'returns the specialization' do
+      expect(@teacher.specialization).to eql 'doctor'
+    end
+  end
```
Comment (danmainah) [inline spec/teacher_spec.rb:31]: - [ ] @Team Great work adding tests for the teacher class, however you forgot to add test for this method  ` can_use_services`, its  project requirement to add tests for all methods
image

File: spec/classroom_spec.rb [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+      expect(classroom.students).to be_empty
+    end
+  end
+
+  describe '#classroom=' do
+    it 'should associate the classroom with the student' do
+      student = double('Student')
+      expect(student).to receive(:classroom=).with(classroom)
+      classroom.add_students(student)
+    end
```
Comment (danmainah) [inline spec/classroom_spec.rb:21]: - [ ] @Team Great work adding tests for the classroom class, however you forgot to add test for this method  ` classroom=`, its  project requirement to add tests for all methods
image

File: spec/rental_spec.rb [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+  describe '#rental_to_json' do
+    it 'returns a hash representing the rental' do
+      expected_hash = {
+        date: @rental_date,
+        person: 'John Doe',
+        book: 'The Great Gatsby'
+      }
+      expect(@rental.rental_to_json).to eql(expected_hash)
+    end
```
Comment (danmainah) [inline spec/rental_spec.rb:31]: - [ ] @Team Great work adding tests for the rental class, however you forgot to add tests for these methods ` self.from_json, save_rental, load_rental`, its  project requirement to add tests for all methods
image

**Review by danmainah — CHANGES_REQUESTED** [review]
Hi @axcell-0 , @BiancaMe and @Omwomo , 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

## Highlights
✔️  Added tests for all the required classes except decoration classes.
✔️  Ensured all tests added are passing.


# Required Changes ♻️

- [x] @Team You forgot to add tests for the classes that you created while implementing the Decorator design pattern, this is part of project requirement
image


_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s).

File: spec/rental_spec.rb
Comment (Efiamotu-1) [inline spec/rental_spec.rb:1]: - [ ] As requested by the previous reviewer in this [review link](https://github.com/Omwomo/school-library/pull/7#discussion_r1380312580), you are yet to add test cases for the `save_rental & load_rental` please kindly do this :+1: :sparkles:

File: spec/person_spec.rb
Comment (Efiamotu-1) [inline spec/person_spec.rb:1]: - [ ] As requested by the previous reviewer in this [review link](https://github.com/Omwomo/school-library/pull/7#discussion_r1380304087) you are yet to add tests for `self.from_json, self.find_by_id,`. Please kindly do this :+1:

File: spec/book_spec.rb
Comment (Efiamotu-1) [inline spec/book_spec.rb:1]: - [ ] As requested by the previous reviewer in this [review link](https://github.com/Omwomo/school-library/pull/7#discussion_r1380295528) you are yet to add test for the method `self.from_json` please kindly add them :+1:

**Review by Efiamotu-1 — CHANGES_REQUESTED** [review]
Hi @Team, 



Great work on making the changes requested by a previous reviewer 👏 Your Code is Awesome! 💪
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.

There are some **few** issues 🐛 that you still need to work 🔧 on though to go to the next project ⏭️ but you are almost there! 🏃
## Don't Worry 🙏

It's ok to have bugs in your code 🐛. Please, see this code review as a **chance** for you to **learn** practical coding skills and learn **new tips** that can improve your code quality 💯 and help you **get a better ** 💼.

_I am going to guide you on how to fix them down below. 🤓_

## Highlights :dart: 
- ✔️ No linter Errors
- ✔️ Descriptive README
- ✔️ Descriptive PR title and description
- The code is top-notch! 💯 and has a huge potential.

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._


## Important Notes 📝

- Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
- Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

> **Note:** Please, don't Merge This Pull Request yet. Only Merge when it is Approved ✅.


## Thank you for your time 🙏 and call me if you need help or clarification ☎️

- This is My **GitHub**: [@Efiamotu-1](https://github.com/Efiamotu-1/)
- This is My **linkedIn**.[Habeeb Efiamotu Musa](https://www.linkedin.com/in/musa-habeeb/)
- This is My **Twitter**: [@EFYAMOTU](https://twitter.com/EFYAMOTU)
- This is My **Slack**: [Habeeb Efiamotu Musa](https://slack.com)

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by Reem-lab — APPROVED** [review]
# Hi Team,  👋🏻👋🏻

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 


## Highlights

- [ ]  Followed Gitflow 💯
- [ ]  Descriptive PR⭐
- [ ]  Kudos for working on the tests🤩
- [ ] All tests passed  🎉
- [ ] Good job working on the requested changes 🙏🏻


## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers, and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**